File: scripts/check-public-contracts/mintlify.go
package main

import (
	"errors"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
	"time"
)

func checkMintlifyAgentReadiness(target string) {
	baseURL := normalizeTargetURL(target)
	retries := envInt("TOVUK_DOCS_CHECK_RETRIES", 8)
	retryDelay := time.Duration(envInt("TOVUK_DOCS_CHECK_RETRY_DELAY_MS", 5000)) * time.Millisecond
	client := &http.Client{Timeout: 20 * time.Second}

	requiredPaths := []string{
		"/llms.txt",
		"/llms-full.txt",
		"/skill.md",
		"/.well-known/skills/index.json",
		"/.well-known/agent-skills/index.json",
		"/.well-known/mcp",
		"/sitemap.xml",
		"/robots.txt",
		"/openapi.json",
	}

	for _, path := range requiredPaths {
		response := fetchText(client, baseURL, path, nil, retries, retryDelay)
		if strings.TrimSpace(response) == "" {
			fail("%s is empty", path)
		}
		rejectRetiredPublicNames(path, response)
	}

	for _, path := range []string{"/", "/quickstart", "/pricing", "/reference/limits"} {
		response := fetchText(client, baseURL, path, map[string]string{"Accept": "text/html"}, retries, retryDelay)
		rejectRetiredPublicNames(path, response)
	}

	llms := fetchText(client, baseURL, "/llms.txt", nil, retries, retryDelay)
	requirePattern("llms.txt", llms, `(?m)^# `)
	requirePattern("llms.txt", llms, `\[[^\]]+\]\([^)]+\)`)

	skill := fetchText(client, baseURL, "/skill.md", nil, retries, retryDelay)
	requirePattern("skill.md", skill, `(?m)^---\n`)
	requirePattern("skill.md", skill, `(?i)name:\s*tovuk`)

	robots := fetchText(client, baseURL, "/robots.txt", nil, retries, retryDelay)
	if regexp.MustCompile(`(?i)Disallow:\s*/`).MatchString(robots) &&
		!regexp.MustCompile(`(?i)Allow:\s*/`).MatchString(robots) {
		fail("robots.txt appears to block crawlers")
	}

	markdown := fetchText(client, baseURL, "/", map[string]string{"Accept": "text/markdown"}, retries, retryDelay)
	requirePattern("Markdown content negotiation", markdown, `Tovuk`)

	plaintext := fetchText(client, baseURL, "/", map[string]string{"Accept": "text/plain"}, retries, retryDelay)
	requirePattern("Plain text content negotiation", plaintext, `Tovuk`)

	mcpDiscovery := fetchText(client, baseURL, "/.well-known/mcp", nil, retries, retryDelay)
	requirePattern("MCP discovery", mcpDiscovery, `"url"\s*:`)
	requirePattern("MCP discovery", mcpDiscovery, `/mcp`)

	fmt.Printf("Mintlify agent readiness checks passed for %s\n", baseURL)
}

func rejectRetiredPublicNames(label string, source string) {
	lower := strings.ToLower(source)
	for _, retired := range retiredPublicNames() {
		if strings.Contains(lower, retired) {
			fail("%s contains retired public branding", label)
		}
	}
	rejectForbiddenPublicCopyTerms(label, source)
}

func retiredPublicNames() []string {
	return []string{
		string([]byte{122, 101, 114, 99, 116}),
		string([]byte{120, 113, 117, 105, 107}),
	}
}

func checkMintlifyScore(path string) {
	var score map[string]interface{}
	readJSON(path, &score)

	value := numberField(score, "score")
	if value == 0 {
		value = numberField(score, "overallScore")
	}
	minimum := float64(envInt("MINTLIFY_SCORE_MIN", 90))
	if value < minimum {
		fail("Mintlify score is %.0f/100; expected at least %.0f/100", value, minimum)
	}
	fmt.Printf("Mintlify score is %.0f/100\n", value)
}

func fetchText(client *http.Client, baseURL string, path string, headers map[string]string, retries int, retryDelay time.Duration) string {
	var lastError error
	for attempt := 0; attempt <= retries; attempt++ {
		text, err := requestText(client, baseURL, path, headers)
		if err == nil {
			return text
		}
		lastError = err
		if attempt == retries || !isRetryableFetchError(err) {
			break
		}
		time.Sleep(retryDelay)
	}
	fail("%s", lastError)
	return ""
}

func requestText(client *http.Client, baseURL string, path string, headers map[string]string) (string, error) {
	request, err := http.NewRequest(http.MethodGet, baseURL+path, nil)
	if err != nil {
		return "", err
	}
	for name, value := range headers {
		request.Header.Set(name, value)
	}

	response, err := client.Do(request)
	if err != nil {
		return "", err
	}
	defer response.Body.Close()

	body, err := io.ReadAll(response.Body)
	if err != nil {
		return "", err
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return "", httpStatusError{Path: path, Status: response.StatusCode}
	}
	return string(body), nil
}

type httpStatusError struct {
	Path   string
	Status int
}

func (errorValue httpStatusError) Error() string {
	return fmt.Sprintf("%s returned %d", errorValue.Path, errorValue.Status)
}

func isRetryableFetchError(err error) bool {
	var statusError httpStatusError
	if errors.As(err, &statusError) {
		return statusError.Status == http.StatusTooManyRequests || statusError.Status >= 500
	}
	return true
}

func normalizeTargetURL(target string) string {
	if !strings.HasPrefix(target, "http://") && !strings.HasPrefix(target, "https://") {
		target = "https://" + target
	}
	return strings.TrimRight(target, "/")
}

File: crates/tovuk/examples/check-public-contracts/mintlify.rs
use std::time::Duration;

use reqwest::{Url, blocking::Client};
use serde_json::Value;

use crate::helpers::{
    CheckResult, env_int, has_markdown_link, number_field, read_json,
    reject_forbidden_public_copy_terms, require_contains, require_contains_all,
    retired_public_names,
};
use crate::mintlify_fetch::{fetch_text, normalize_target_url, retry_delay};

pub(crate) fn check_agent_readiness(target: &str) -> CheckResult {
    let base_url = normalize_target_url(target);
    let retries = env_int("TOVUK_DOCS_CHECK_RETRIES", 8)?;
    let retry_delay = retry_delay()?;
    let client = Client::builder()
        .timeout(Duration::from_secs(20))
        .build()
        .map_err(|error| format!("build HTTP client: {error}"))?;

    check_required_agent_paths(&client, base_url.as_str(), retries, retry_delay)?;
    check_html_paths(&client, base_url.as_str(), retries, retry_delay)?;
    check_llms_skill_and_robots(&client, base_url.as_str(), retries, retry_delay)?;
    check_content_negotiation(&client, base_url.as_str(), retries, retry_delay)?;
    check_mcp_discovery(&client, base_url.as_str(), retries, retry_delay)?;

    println!("Mintlify agent readiness checks passed for {base_url}");
    Ok(())
}

pub(crate) fn check_score(path: &str) -> CheckResult {
    let score: Value = read_json(path)?;
    let mut value = number_field(&score, "score");
    if value == 0.0 {
        value = number_field(&score, "overallScore");
    }
    let minimum = f64::from(
        i32::try_from(env_int("MINTLIFY_SCORE_MIN", 90)?)
            .map_err(|_| "MINTLIFY_SCORE_MIN must fit in an i32".to_owned())?,
    );
    if value < minimum {
        return Err(format!(
            "Mintlify score is {value:.0}/100; expected at least {minimum:.0}/100"
        ));
    }
    println!("Mintlify score is {value:.0}/100");
    Ok(())
}

fn check_required_agent_paths(
    client: &Client,
    base_url: &str,
    retries: i64,
    retry_delay: Duration,
) -> CheckResult {
    for path in [
        "/llms.txt",
        "/llms-full.txt",
        "/skill.md",
        "/.well-known/skills/index.json",
        "/.well-known/agent-skills/index.json",
        "/.well-known/mcp",
        "/sitemap.xml",
        "/robots.txt",
        "/openapi.json",
    ] {
        let response = fetch_text(client, base_url, path, &[], retries, retry_delay)?;
        if response.trim().is_empty() {
            return Err(format!("{path} is empty"));
        }
        reject_retired_public_names(path, response.as_str())?;
    }
    Ok(())
}

fn check_html_paths(
    client: &Client,
    base_url: &str,
    retries: i64,
    retry_delay: Duration,
) -> CheckResult {
    for path in ["/", "/quickstart", "/pricing", "/reference/limits"] {
        let response = fetch_text(
            client,
            base_url,
            path,
            &[("Accept", "text/html")],
            retries,
            retry_delay,
        )?;
        reject_retired_public_names(path, response.as_str())?;
    }
    Ok(())
}

fn check_llms_skill_and_robots(
    client: &Client,
    base_url: &str,
    retries: i64,
    retry_delay: Duration,
) -> CheckResult {
    let llms = fetch_text(client, base_url, "/llms.txt", &[], retries, retry_delay)?;
    if !llms.lines().any(|line| line.starts_with("# ")) {
        return Err("llms.txt did not include a Markdown heading".to_owned());
    }
    if !has_markdown_link(llms.as_str()) {
        return Err("llms.txt did not include a Markdown link".to_owned());
    }

    let skill = fetch_text(client, base_url, "/skill.md", &[], retries, retry_delay)?;
    if !skill.starts_with("---\n") {
        return Err("skill.md did not include frontmatter".to_owned());
    }
    if !skill.lines().any(|line| {
        line.to_lowercase().starts_with("name:") && line.to_lowercase().contains("tovuk")
    }) {
        return Err("skill.md did not include name: tovuk".to_owned());
    }

    let robots = fetch_text(client, base_url, "/robots.txt", &[], retries, retry_delay)?;
    if robots_blocks_crawlers(robots.as_str()) {
        return Err("robots.txt appears to block crawlers".to_owned());
    }
    Ok(())
}

fn check_content_negotiation(
    client: &Client,
    base_url: &str,
    retries: i64,
    retry_delay: Duration,
) -> CheckResult {
    let markdown = fetch_text(
        client,
        base_url,
        "/",
        &[("Accept", "text/markdown")],
        retries,
        retry_delay,
    )?;
    require_contains(markdown.as_str(), "Tovuk", "Markdown content negotiation")?;

    let plaintext = fetch_text(
        client,
        base_url,
        "/",
        &[("Accept", "text/plain")],
        retries,
        retry_delay,
    )?;
    require_contains(
        plaintext.as_str(),
        "Tovuk",
        "Plain text content negotiation",
    )
}

fn check_mcp_discovery(
    client: &Client,
    base_url: &str,
    retries: i64,
    retry_delay: Duration,
) -> CheckResult {
    let mcp_discovery = fetch_text(
        client,
        base_url,
        "/.well-known/mcp",
        &[],
        retries,
        retry_delay,
    )?;
    require_contains_all(
        mcp_discovery.as_str(),
        &[
            (r#""url""#, "MCP discovery"),
            (":", "MCP discovery"),
            ("/mcp", "MCP discovery"),
        ],
    )?;
    require_mcp_urls_on_base_host(base_url, mcp_discovery.as_str())
}

fn require_mcp_urls_on_base_host(base_url: &str, source: &str) -> CheckResult {
    let base = Url::parse(base_url).map_err(|error| format!("parse docs base URL: {error}"))?;
    let base_host = base
        .host_str()
        .ok_or_else(|| format!("docs base URL must include a host: {base_url}"))?;
    let discovery = serde_json::from_str::<Value>(source)
        .map_err(|error| format!("parse MCP discovery JSON: {error}"))?;
    let mut urls = Vec::new();
    collect_url_fields(&discovery, &mut urls);
    if urls.is_empty() {
        return Err("MCP discovery did not include URL fields".to_owned());
    }
    for url in urls {
        require_mcp_url_on_base_host(base.scheme(), base_host, url)?;
    }
    Ok(())
}

fn collect_url_fields<'a>(value: &'a Value, urls: &mut Vec<&'a str>) {
    match value {
        Value::Array(items) => {
            for item in items {
                collect_url_fields(item, urls);
            }
        }
        Value::Object(fields) => {
            for (name, item) in fields {
                if name == "url" {
                    if let Value::String(url) = item {
                        urls.push(url.as_str());
                    }
                }
                collect_url_fields(item, urls);
            }
        }
        _ => {}
    }
}

fn require_mcp_url_on_base_host(base_scheme: &str, base_host: &str, url: &str) -> CheckResult {
    let parsed =
        Url::parse(url).map_err(|error| format!("parse MCP discovery URL {url}: {error}"))?;
    if parsed.scheme() != base_scheme || parsed.host_str() != Some(base_host) {
        return Err(format!(
            "MCP discovery URL {url} must stay on {base_scheme}://{base_host}"
        ));
    }
    if parsed.path() != "/mcp" {
        return Err(format!(
            "MCP discovery URL {url} must use the public /mcp path"
        ));
    }
    Ok(())
}

fn reject_retired_public_names(label: &str, source: &str) -> CheckResult {
    let lower = source.to_lowercase();
    for retired in retired_public_names() {
        if lower.contains(retired) {
            return Err(format!("{label} contains retired public branding"));
        }
    }
    reject_forbidden_public_copy_terms(label, source)
}

fn robots_blocks_crawlers(source: &str) -> bool {
    let disallows_all = source.lines().any(|line| {
        let lower = line.to_lowercase();
        lower
            .split_once(':')
            .filter(|(name, _)| name.trim() == "disallow")
            .is_some_and(|(_, value)| value.trim_start().starts_with('/'))
    });
    let allows_all = source
        .lines()
        .any(|line| line.to_lowercase().trim() == "allow: /");
    disallows_all && !allows_all
}

#[cfg(test)]
mod tests {
    use super::require_mcp_urls_on_base_host;

    #[test]
    fn accepts_public_mcp_urls_on_docs_host() {
        let source = r#"{
            "url": "https://docs.tovuk.com/mcp",
            "servers": [
                {"name": "public", "url": "https://docs.tovuk.com/mcp"}
            ]
        }"#;

        assert!(
            require_mcp_urls_on_base_host("https://docs.tovuk.com", source).is_ok(),
            "MCP discovery should accept public docs-domain URLs"
        );
    }

    #[test]
    fn rejects_generated_mintlify_hosts() {
        let source = r#"{
            "url": "https://project-123.mintlify.me/mcp",
            "servers": [
                {"name": "public", "url": "https://project-123.mintlify.me/mcp"}
            ]
        }"#;

        let result = require_mcp_urls_on_base_host("https://docs.tovuk.com", source);
        assert!(
            matches!(result, Err(message) if message.contains("must stay on https://docs.tovuk.com")),
            "MCP discovery should reject generated Mintlify hosts"
        );
    }

    #[test]
    fn rejects_non_public_mcp_paths() {
        let source = r#"{
            "url": "https://docs.tovuk.com/authed/mcp",
            "servers": [
                {"name": "public", "url": "https://docs.tovuk.com/authed/mcp"}
            ]
        }"#;

        let result = require_mcp_urls_on_base_host("https://docs.tovuk.com", source);
        assert!(
            matches!(result, Err(message) if message.contains("public /mcp path")),
            "MCP discovery should use the public /mcp path"
        );
    }
}
